Allow arbitrary injection into middleware and param converters

felixfbecker opened the pull request:
This is an implementation for #15
I would love to get this merged as quickly as possible to depend on it in my projects.

---
### Feature

It is now possible to inject whatever is registered in the container into route param converters and middleware via type hints. Also, as long as the arguments are type-hinted, argument position of request / response / app doesn't matter anymore. If there are no type-hints, it falls back to Silex' default argument order.
### Implementation

To accomplish this, `ConverterListener` and `MiddlewareListener` are overwritten and in `Application` the default dispatcher is overwritten to use the custom classes. For application middleware, `before`/`after`/`finish` are overwritten aswell.

For DRY reasons the injection logic is done in separate classes (`BeforeMiddlewareCaller`, `AfterMiddlewareCaller`, `ConverterCaller`). This allows the application middleware and route middleware to share the same logic and also for all event listeners to share the same instance of the caller and `ParameterResolver` objects.
### Tests

`ConverterTest` and `MiddlewareTest` test that both arbitrary injection is possible and the old behaviour still works.
### Docs

I altered the README to highlight that injection works not only for controllers but also for converters and middleware and added two examples.
### BC Breaks

None

---

Any thoughts / objections?

felixfbecker committed:
Allow arbitrary injection into middleware and param converters

mnapoli commented on src/MiddlewareListener.php:
```diff
@@ -0,0 +1,77 @@
+ *
+ * @author Felix Becker <f.becker@outlook.com>
+ */
+class MiddlewareListener extends \Silex\EventListener\MiddlewareListener
```
I'm not sure I understand when this class is used by Silex? Since `App::before()` and `App::after()` are overridden, why do we need to override this too?

felixfbecker replied:
There are two types of middleware, Application level middleware via `$app->before()` and route / controller collection level middleware like `$app->get('/', ...)->before(...)`. Route level middleware is handled throught `MiddlewareListener`.

mnapoli replied:
Ah right thanks!

mnapoli commented on src/AfterMiddlewareCaller.php:
```diff
@@ -0,0 +1,90 @@
+            ));
+        }
+
+        return call_user_func_array($middleware, $arguments);
```
You could use [`Invoker::call()`](https://github.com/PHP-DI/Invoker#usage) instead of reimplementing all the behavior?

felixfbecker replied:
I took the implementation from [`ControllerResolver::getArguments`](https://github.com/PHP-DI/Silex-Bridge/blob/master/src/Controller/ControllerResolver.php#L65). Does `Invoker::call()` support a custom `ResolverChain`? Because I have to fallback to a `NumericParameterResolver`

felixfbecker replied:
It is also important to resolve the request object at any parameter position before calling `getParameters` because otherwise PHPDI will create a new request object if it sees the `Request` type hint

mnapoli commented on src/BeforeMiddlewareCaller.php:
```diff
@@ -0,0 +1,85 @@
+ *
+ * @author Felix Becker <f.becker@outlook.com>
+ */
+class BeforeMiddlewareCaller
```
Are there any differences between `BeforeMiddlewareCaller` and `AfterMiddlewareCaller`? Maybe there can be only one class?

felixfbecker replied:
The difference is that after middleware gets passed `($request, $response, $app)` while before middleware gets passed `($request, $app)`, because the response doesn't exist yet. I feel like making them inherit from an abstract base class and refactoring few lines of code into common methods would result in more code than it would save.

felixfbecker commented on src/Application.php:
```diff
@@ -100,4 +159,57 @@ public function getPhpDi()
+
+    public function before($callback, $priority = 0)
+    {
+        $app = $this;
```
just noticed this isn't needed, will remove later

felixfbecker commented:
I will squash once you give your OK to merge

jdreesen commented on src/AfterMiddlewareCaller.php:
```diff
@@ -0,0 +1,90 @@
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Class that takes the responsibility of resolving and calling a before middleware with injection
```
"calling ~~a before~~ **an after** middleware"

jdreesen commented on src/Application.php:
```diff
@@ -70,6 +102,33 @@ public function __construct(ContainerBuilder $containerBuilder = null, array $va
+
+        // Override the dispatcher with ours to use our event listeners
+        $this['dispatcher'] = $this->share(function () {
+            /*
```
Missing `*`

felixfbecker replied:
Funny, this is copied directly from the Silex source

jdreesen commented on tests/ConverterTest.php:
```diff
@@ -0,0 +1,81 @@
+
+        $request = Request::create('/john?some=param');
+
+        $test = $this;
```
This should not be neccessary because PHP 5.4 supports `$this` in closures.

felixfbecker replied:
I wasn't sure about this, but Silex does this everywhere.

jdreesen replied:
That's because Silex 1 allows PHP >= 5.3 but this bridge is PHP >= 5.4 only.

felixfbecker replied:
ah, thanks. I do a lot of JS development so I wasn't even sure about how PHP resolves $this in closures.

jdreesen commented on src/MiddlewareListener.php:
```diff
@@ -0,0 +1,77 @@
+        foreach ((array) $route->getOption('_before_middlewares') as $callback) {
+
+            $ret = $this->beforeMiddlewareCaller->call($callback, $request);
+
```
Blank lines should be aligned with those of the `foreach` in `onKernelResponse()`.

jdreesen commented on src/MiddlewareListener.php:
```diff
@@ -0,0 +1,77 @@
+
+            if ($ret instanceof Response) {
+                $event->setResponse($ret);
+                return;
```
Unnecessary `return`

jdreesen commented:
As I don't have much time right now I just took a quick look.
The PR looks good overall, just some small nits. 
Although I don't really like the code duplications in all those `call()` methods...

felixfbecker commented:
@jdreesen I addressed your comments and squashed.

> Although I don't really like the code duplications in all those call() methods...

I know, I don't like it either, but every case (after middleware, before middleware, param converter, controller resolver) is just a bit different so they cannot share the exact same implementation. If you break it down it's really nothing but
- do reflection (by using the already high-level `Invoker` API)
- do a `foreach` to resolve the very special arguments needed in that case (request, response etc)
- get the arguments with fallback to the default behaviour in that case
- check if all parameters resolved by getting a diff and maybe throwing an exception.  
  this is the only step (4 statements if you count) that all cases actually share and it doesn't make sense to refactor this anywhere

I'm actually quite happy that there is no duplication for route / application level middleware and that finish and after middleware can both use `AfterMiddlewareCaller`.

mnapoli commented:
Sorry for the delay, I've taken some time to think about it.

In order to preserve BC we want to inject the request, response and app by:
- parameter name
- type-hint

The application is not a problem because it's a service, but the request & response are not so using the Invoker, even with a custom `ResolverChain` (to answer your question yes it's possible), will not work because type-hints are only checked for container entries.

What we could do is create a new type of `ParameterResolver` that checks the type-hint, but not for container entries. Example of usage:

``` php
$parameterResolver = new ResolverChain([
    new AssociativeArrayResolver, // inject request/response by name
    new TypeHintResolver, // inject request/response by type-hint
    new TypeHintContainerResolver, // dependency injection
]);

...

$invoker->call($middleware, [
    'request' => $request, // by name, used by AssociativeArrayResolver
    'response' => $response, // by name, used by AssociativeArrayResolver
    'Symfony\...\Request' => $request, // by type-hint, used by TypeHintResolver
    'Symfony\...\Response' => $request, // by type-hint, used by TypeHintResolver
]);
```

The `TypeHintResolver` could be reused by other bridges later (e.g. Slim framework). It could be added directly to the Invoker project.

What do you think?

felixfbecker commented:
@mnapoli 

> In order to preserve BC we want to inject the request, response and app by:
> - parameter name

that doesn't make much sense to me - with default Silex there is not injection by parameter name, only by parameter position. A middleware is always called with `call_user_func($middleware, $request, $app)`. I took care of BC by falling back to a numeric parameter resolver. I get injection by parameter name for route parameters, but injecting the request by parameter name is too magic for me. It would also be quite weird to document because for consistency, _all_ parameters including the application should then be able to get injected by name, and the application is always called `$app` in the docs...

> What we could do is create a new type of ParameterResolver that checks the type-hint, but not for container entries

this is a very good solution, but I want to point out here that I strictly followed the existing implementation in `ControllerResolver`, so this has not much to do with this PR. But it would save a lot of code duplication.

What I don't get though:

> ``` php
> $invoker->call($middleware, [
>    'request' => $request, // by name, used by AssociativeArrayResolver
>    'response' => $response, // by name, used by AssociativeArrayResolver
>    'Symfony\...\Request' => $request, // by type-hint, used by TypeHintResolver
>    'Symfony\...\Response' => $request, // by type-hint, used by TypeHintResolver
> ]);
> ```

Where do you pass the paramResolver here?

mnapoli commented:
The parameter resolvers are passed when constructing the Invoker:

``` php
$parameterResolver = new ResolverChain([
    ...
]);
$invoker = new Invoker\Invoker($parameterResolver);
```

And you are right about injection of request/response by name, that's not necessary here so let's keep it as simple as we can (and it wouldn't be consistent with injection in controllers anyway).

Then you would only need something like this:

``` php
$parameterResolver = new ResolverChain([
    new TypeHintResolver, // inject request/response by type-hint
    new TypeHintContainerResolver, // dependency injection
    new NumericArrayResolver, // fallback to injecting "request, response, application"
]);
$invoker = new Invoker\Invoker($parameterResolver);

$invoker->call($middleware, [
    // parameters indexed by position for the NumericArrayResolver
    $request, $response, $application,
    // parameters indexed by the type-hint for the TypeHintResolver
    'Symfony\...\Request' => $request,
    'Symfony\...\Response' => $request,
]);
```

Does that make sense?

> I want to point out here that I strictly followed the existing implementation in ControllerResolver, so this has not much to do with this PR

I agree and take your point. But with this PR this implementation gets duplicated multiple times and the code gets much more complex. By merging this I'll also be committing to supporting this code, I'd rather go with something as simple as possible ;)

felixfbecker commented:
@mnapoli Your proposed solution makes a lot of sense. Do you want me to do a PR to Invoker?

mnapoli commented:
@felixfbecker if you have time for that that would be awesome!

felixfbecker commented:
I changed the implementation to use the `TypeHintResolver` from my PR https://github.com/PHP-DI/Invoker/pull/8.
- A single class `CallbackInvoker` does the job of all the `*Caller` classes before
- It extends `Invoker` for type safety so `MiddlewareListener` etc. can depend on this specific `Invoker` with this specific `ResolverChain`.
- Tests of course fail because you need to copy the new `TypeHintResolver` to vendor

When the change is published on Invoker I will bump the dependency version and squash. Just wanted to share the changes now so you can review.

felixfbecker commented:
Looks good?

mnapoli commented on src/CallbackInvoker.php:
```diff
@@ -0,0 +1,34 @@
+            new TypeHintContainerResolver($container),
+            new NumericArrayResolver,
+        ]));
+    }
```
I'm not convinced by the usefulness of this class. I think you referred to that when you said:

> It extends Invoker for type safety so MiddlewareListener etc. can depend on this specific Invoker with this specific ResolverChain.

But I don't see why a MiddlewareListener should explicitly require this implementation. To me it would be better than middleware listeners or parameter converters are implementations that rely on the interface of "Invoker", and then the application passes an instance of `Invoker` configured as we want.

I.e. I don't see the point to couple things more than necessary.

felixfbecker replied:
But then the application could theoretically pass an `Invoker` instance that uses a completely different `ResolverChain`, which would break the `MiddlewareListener`. Only this way can `MiddlewareListener` declare that it needs an `Invoker` that is suited for calling Silex callbacks, through the parameter names, type hints, the container and through parameter position.

felixfbecker replied:
I am not attached to it though. If you feel strongly about this, let me know and I remove the class :)

mnapoli commented on src/Application.php:
```diff
@@ -70,6 +90,33 @@ public function __construct(ContainerBuilder $containerBuilder = null, array $va
+            $dispatcher->addSubscriber(new StringToResponseListener());
+
+            return $dispatcher;
+        });
```
I feel uneasy copying all that configuration code from Silex :/ If they add something, we will fall behind for sure (I know I wont do a diff regularly, that's too much work).

Here is an idea: extend the container entry (`$container->extend(...)`) to remove the previous MiddlewareListener and replace it by our own? I don't know if we can easily browse the registered event subscribers though… Damn Silex is so not extensible :(

felixfbecker replied:
I feel the same. I tried to implement this by getting all listeners with `getListeners()` and iterating through them, to find the Silex middleware subscriber, remove that and add our own. But that was more code in the end, and less performant for sure. I also don't know if the changed order of subscribers would cause any problem. It is unfortunate, but this is the problem everytime you overwrite something in OOP, you always duplicate some implementation.

mnapoli commented:
Looking very good! Much better now! It's really too bad we have to copy so much code from Silex :/ but that's not your fault of course.

felixfbecker commented:
Do you want me to squash so you can merge? Or do you have any other idea for the dispatcher?

mnapoli commented:
Thanks for your answers. I'm good for merging, :+1: for squashing

felixfbecker commented:
:+1:

felixfbecker commented:
It should be good to go, any ETA?

mnapoli commented:
A few minutes :)

mnapoli commented:
[1.5.0](https://github.com/PHP-DI/Silex-Bridge/releases/tag/1.5.0) is tagged and released! The documentation was also updated in PHP-DI's docs: http://php-di.org/doc/frameworks/silex.html

Thank you for your patience and implementing all that ;)
